flink-ml-lib/src/main/java/org/apache/flink/ml/classification/linear/LogisticRegression.java

```diff
@@ -0,0 +1,653 @@
+import org.apache.flink.streaming.api.operators.TwoInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.OutputTag;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.commons.collections.IteratorUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+/**
+ * This class implements methods to train a logistic regression model. For details, see
+ * https://en.wikipedia.org/wiki/Logistic_regression.
+ */
+public class LogisticRegression
+        implements Estimator<LogisticRegression, LogisticRegressionModel>,
+                LogisticRegressionParams<LogisticRegression> {
+
+    private Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    private static final OutputTag<Tuple2<double[], double[]>> MODEL_OUTPUT =
+            new OutputTag<Tuple2<double[], double[]>>("MODEL_OUTPUT") {};
+
+    public LogisticRegression() {
+        ParamUtils.initializeMapWithDefaultValues(this.paramMap, this);
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static LogisticRegression load(StreamExecutionEnvironment env, String path)
+            throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public LogisticRegressionModel fit(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+
+        DataStream<Tuple3<Double, Double, double[]>> trainData =
+                tEnv.toDataStream(inputs[0])
+                        .map(
+                                dataPoint ->
+                                        Tuple3.of(
+                                                getWeightCol() == null
+                                                        ? new Double(1.0)
+                                                        : (Double)
+                                                                dataPoint.getField(getWeightCol()),
+                                                (Double) dataPoint.getField(getLabelCol()),
+                                                (double[]) dataPoint.getField(getFeaturesCol())))
+                        .returns(
+                                new TupleTypeInfo<>(
```

Comment:
Many algorithms have `features`, `weight` and `label` and it is common for those algorithms to first convert the input `Row` into `Tuple3` of these fields. Hopefully we can setup a best practice of the order of these fields in the Tuple, so that the algorithm logic could consistent and intuitive to read.

My understanding is that input (i.e. `features`, `weight`) usually comes before the expected output (i.e. `label`). And it is more intuitive for `features` to be before `weight` because `features` are more important. So the Tuple could be `<features, weight, label>`. What do you think?

Reply:
And can we also put comments to explain the semantic meaning of fields of `trainData`?

Reply:
Shall we add a class to represent one data point? e.g., 

`class Instance(label: Double, weight: Double, features: DenseVector)`

Reply:
IMO the point might be a bit overkill in this case. If we choose to a dedicated class instead of a Tuple, I worry we might either end up with many classes due to the possible combination of fields (see NaiveBayes for example), or end up with inconsistent code.

For now it may be better to use Tuple and make the order of fields consistent with most developers' intuition and other algorithms' implementation.

Reply:
As discussed offline, we decided to go with a specific class `LabeledPointWithWeight `.